execution-engine/src/wasi/wasmtime.rs

```diff
@@ -70,84 +73,106 @@ macro_rules! convert_wasi_arg {
         // versions of wasmtime
-        &mut (unsafe { self.memory.data_unchecked_mut() })[self.address..self.address + self.length]
+        &mut (self.memory.data_mut(&mut self.store))[self.address..self.address + self.length]
     }
 }
 
-impl MemorySlice for WasmtimeSlice {}
-impl MemorySliceMut for WasmtimeSlice {}
+impl<T> MemorySlice for WasmtimeSlice<'_, T> {}
+//TODO: REMOVE??
+//impl<T>  MemorySlice for WasmtimeSliceMut<'_, T> {}
+impl<T> MemorySliceMut for WasmtimeSliceMut<'_, T> {}
 
 /// Impl the MemoryHandler for Caller.
 /// This allows passing the Caller to WasiWrapper on any VFS call.
-impl MemoryHandler for Caller<'_> {
-    type Slice = WasmtimeSlice;
-    type SliceMut = WasmtimeSlice;
+impl<'a, T: 'static> MemoryHandler for Caller<'a, T> {
+    type Slice = WasmtimeSlice<'static, T>;
+    type SliceMut = WasmtimeSliceMut<'static, T>;
 
-    fn get_slice<'a>(
-        &'a self,
+    fn get_slice<'b>(
+        &'b self,
         address: u32,
         length: u32,
-    ) -> FileSystemResult<Bound<'a, Self::Slice>> {
-        let memory = match self
+    ) -> FileSystemResult<Bound<'b, Self::Slice>> {
+        // NOTE: manually and temporarily change the mutability.
+        let memory = match unsafe { (self as *const Self as *mut Self).as_mut() }
+            .unwrap()
             .get_export(WasiWrapper::LINEAR_MEMORY_NAME)
             .and_then(|export| export.into_memory())
         {
             Some(s) => s,
             None => return Err(ErrNo::NoMem),
         };
         Ok(Bound::new(WasmtimeSlice {
+            store: unsafe {
+                mem::transmute::<StoreContext<'b, T>, StoreContext<'static, T>>(self.as_context())
+            },
             memory,
             address: address as usize,
             length: length as usize,
         }))
     }
 
-    fn get_slice_mut<'a>(
-        &'a mut self,
+    fn get_slice_mut<'c>(
+        &'c mut self,
         address: u32,
         length: u32,
-    ) -> FileSystemResult<BoundMut<'a, Self::SliceMut>> {
+    ) -> FileSystemResult<BoundMut<'c, Self::SliceMut>> {
         let memory = match self
             .get_export(WasiWrapper::LINEAR_MEMORY_NAME)
             .and_then(|export| export.into_memory())
         {
             Some(s) => s,
             None => return Err(ErrNo::NoMem),
         };
-        Ok(BoundMut::new(WasmtimeSlice {
+        Ok(BoundMut::new(WasmtimeSliceMut {
+            store: unsafe {
+                mem::transmute::<StoreContextMut<'c, T>, StoreContextMut<'static, T>>(
+                    self.as_context_mut(),
+                )
+            },
             memory,
             address: address as usize,
             length: length as usize,
         }))
     }
 
     fn get_size(&self) -> FileSystemResult<u32> {
-        let memory = match self
+        // NOTE: manually and temporarily change the mutability.
+        let memory = match unsafe { (self as *const Self as *mut Self).as_mut() }
+            .unwrap()
```

Comment:
When can this fail?

Reply:
It only fails if the raw pointer is null. It never happens here.